### ISSUE TITLE
Proposal: Roughtime container

### ISSUE THREAD

**Issue by riyazdf** [linuxkit/linuxkit#1699]
We should consider including a Roughtime package as an authenticated and misbehavior-resistant replacement for NTP: https://roughtime.googlesource.com/roughtime/

**Comment by riyazdf** [comment]
cc @avsm @samoht @yomimono in case this can somehow be combined with the mirages SDK work, I know NTP was on the shortlist of next implementations

**Comment by justincormack** [comment]
It is useful for some things, but bear in mind it is not precise enough for `etcd` so not suitable for Docker or Kubernetes if it is 10s ish accuracy...

**Comment by riyazdf** [comment]
yeah, ~10s is too long for raft as well since I think it's 2s heartbeats? cc @diogomonica

**Comment by diogomonica** [comment]
@riyazdf raft uses counters, I'm not sure if the timestamps have any precision requirements there.

**Comment by avsm** [comment]
The other simpler alternative for one-shot time is https://github.com/ioerror/tlsdate as well, to put that on the list

**Comment by justincormack** [comment]
Docker Swarm mode will fail if the timestamps are more than 2s off. I know because I was debugging some issues related to that. Its the `etcd` implementation that requires that.

**Comment by diogomonica** [comment]
@justincormack only for managers participating in raft. Arguably this is more of a worker-specific feature since we could have worker diversity (multiple OSes, etc). One idea could be workers syncing time from the managers using something secure.

**Comment by justincormack** [comment]
hmm, at present the Docker Editions don't use different images for workers and managers, but of course they could.
